src/bestfit.1
.\"                                      Hey, EMACS: -*- nroff -*-
.\" First parameter, NAME, should be all caps
.\" Second parameter, SECTION, should be 1-8, maybe w/ subsection
.\" other parameters are allowed: see man(7), man(1)
.TH bestfit 1 "April 23, 2001"
.\" Please adjust this date whenever revising the manpage.
.\"
.\" Some roff macros, for reference:
.\" .nh        disable hyphenation
.\" .hy        enable hyphenation
.\" .ad l      left justify
.\" .ad b      justify to both left and right margins
.\" .nf        disable filling
.\" .fi        enable filling
.\" .br        insert line break
.\" .sp <n>    insert n+1 empty lines
.\" for manpage-specific macros, see man(7)
.SH NAME
bestfit \- Optimally choose files to be put on a CD (or other media)
.SH SYNOPSIS
.B bestfit
.RI [ OPTION ]... " FILE"... " " [ "-e COMMAND"... ]
.SH DESCRIPTION
This manual page document describes the \fBbestfit\fP command.
Bestfit is a small program to determine which files that should be put
on a CD (or other media), so that as little space as possible is wasted.
It is very easy to use: you specify files on the command line, and bestfit
prints the names of those that were selected. Alternatively, bestfit
can execute a command for each selected file (e.g. to move them to a
different directory).

This problem is also known as the 0-1 knapsack problem. The algorithm
implemented by bestfit solves instances of it optimally in theta(n*W)
time, where n is number of files to choose between and W the amount of
free space. This may sound bad but since W is number of blocks and not
bytes, this algorithm is not that inefficient. Try it yourself and see.

Due to the nature of this algorithm, bestfit uses a lot of memory -
approximately 1.5M per file specified on command line. As long as
you have enough swap space, this is usually not a problem since
bestfit doesn't use all at the same time.

Bestfit does (at this time) not use any external program such as mkisofs
or mkhybrid to calculate the size of files. Instead the file and directory
sizes and rounded up towards the nearest 2048-byte multiple and added
together. (2048 bytes is the block size on ISO images.)

Bestfit can not account for all oddities with ISO images (for instance,
there are multiple path tables, padding, etc which is not easy to
calculate), therefore exact results can not always be produced. Bestfit
may waste more space than necessary, or even tell you that files that
won't fit actually fits. (I consider the latter to be a bug though.
If you encounter it, please mail me.) In other words, manually selecting
files and checking that they fit using mkisofs may in some cases give
you better results than using bestfit.
.SH OPTIONS
These programs follow the usual GNU command line syntax, with long
options starting with two dashes (`-').
A summary of options is included below.
.TP
.B \-8, \-\-80min
Space is that of an empty 80 min CD. This is the default.
.TP
.B \-7, \-\-74min
Space is that of an empty 74 min CD.
.TP
.B \-d, \-\-dirs-only
Ignore all non-directory arguments. (This is useful when including
files using wildcard/glob patterns.)
.TP
.B \-s, \-\-space=BLOCKS
The maximum amount of space allowed to use, in blocks.
.TP
.B \-r, \-\-reserve=BLOCKS
The number of blocks to reserve (usually 200 by default). This
value is simple substracted from the space value before the
algorithm is started.
.TP
.B \-e, \-\-execute COMMAND...
Execute a command for each selected file, instead of printing
the name on screen. All following arguments after --execute (or -e) are
taken to be arguments to the command to execute for each selected
file. The string `{}' is replaced by the current file name being
processed, and can be used multiple times.

If --invert is used as well as --execute, the command is executed
for files that weren't selected instead.
.TP
.B \-i, \-\-invert
Print names of those files that weren't selected (i.e. did not fitting)
instead.
.TP
.B \-v, \-\-verbose
Explain what is being done.
.TP
.B \-q, \-\-quiet
Output no diagnostic/information messages. (Error messages are still
printed.)
.TP
.B \-h, \-\-help
Show summary of options.
.TP
.B \-v, \-\-version
Output version information and exit.
.SH EXAMPLES
Assume we want burn the following directories on a CD:

  /store/dir1	40M
  /store/dir2	67M
  /store/dir3	38M
  /store/dir4	112M
  /store/dir5	130M
  /store/dir6	82M
  /store/dir7	240M
  /store/dir8	56M
  /store/dir9	94M

These files won't fit on an 80 min CD so we need to burn them on two.
We can have bestfit determine which files to burn on the first CD:

  $ bestfit /store/dir* 
  /store/dir2
  /store/dir4
  /store/dir5
  /store/dir7
  /store/dir8
  /store/dir9
  used 357888 of 439808 blocks (699.0M of 859.0M)
  wasted 1912 of 359800 blocks (3.7M of 702.7M)

Now maybe we want to move these files to another directory. bestfit
can do this too:

  $ bestfit /store/dir* -e mv -i {} /store/to-burn/
  used 357888 of 439808 blocks (699.0M of 859.0M)
  wasted 1912 of 359800 blocks (3.7M of 702.7M)
  $ ls /store/to-burn/foo
  /store/dir2
  /store/dir4
  /store/dir5
  /store/dir7
  /store/dir8
  /store/dir9

The -i option was added to make move interactive, so that
we won't overwrite files by mistake. Note that we don't need
to terminate bestfit's options using --, because -e stops
bestfit from reading further options.
.SH AUTHOR
bestfit was written by Oskar Liljeblad <oskar@osk.mine.nu>.


NEWS
2001-10-01

The total block variable is now 64 bits, fixing possible
overflow problems.
The author has changed email-address to oskar@osk.mine.nu.
Added patch from Steven Enderle to add support for FreeBSD.
Updated to automake 1.4p4.

2001-04-23

First public release (0.1.0).

-


src/bestfit.c
/* bestfit.c - The main source file
 * 
 * Copyright (C) 1998-2001  Oskar Liljeblad
 *
 * This file is part of bestfit.
 *
 * This software is copyrighted work licensed under the terms of the
 * GNU General Public License. Please consult the file "LICENSE" for
 * details.
 */

#include "bestfit.h"

#define _GNU_SOURCE

#ifdef HAVE_SYS_TYPES_H
#include <sys/types.h>
#endif
#ifdef HAVE_SYS_WAIT_H
#include <sys/wait.h>
#endif
#ifdef HAVE_UNISTD_H
#include <unistd.h>
#endif
#ifdef HAVE_LIMITS_H
#include <limits.h>
#endif
#ifdef HAVE_ERRNO_H
#include <errno.h>
#endif
#ifdef STDC_HEADERS
#include <stdarg.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#endif

#include "getopt.h"

/* Verbosity level: 0 = quiet, 1 = normal, 2 = verbose */
static int verbosity = 1;
/* Program name: same as argv[0] */
static char *program_name;

/* VERSION_MESSAGE:
 * Message to display when the --version option is used.
 */
#define VERSION_MESSAGE \
PACKAGE " " VERSION "\n\
Written by Oskar Liljeblad.\n\
\n\
Copyright (C) 2001 Oskar Liljeblad\n\
This is free software; see the source for copying conditions.  There is NO\n\
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.\n"

/* HELP_MESSAGE:
 * Message to display when the --help option is used.
 */
#define HELP_MESSAGE "\
Usage: %s [OPTION]... FILE... [-e COMMAND...]\n\
Determine which FILEs should be put on a CD (or other disk) so that as\n\
little space as possible is wasted. Output the name of those files,\n\
or if --execute specified, execute a command for each.\n\
\n\
  -8, --80min               space is that of an empty 80 min CD (default)\n\
  -7, --74min               space is that of an empty 74 min CD\n\
  -5, --sldvd               space is that of an empty 4,7 GB DVD \n\
  -9, --dldvd               space is that of an empty 8,5 GB DVD \n\
  -d, --dirs-only           ignore non-directory arguments\n\
  -s, --space=BLOCKS        maximum amount of space allowed to use\n\
  -r, --reserve=BLOCKS      blocks to reserve (%d by default)\n\
  -e, --execute COMMAND...  execute commands for each selected file\n\
  -i, --invert              print/execute for non-fitting files instead\n\
  -v, --verbose             explain what is being done\n\
  -q, --quiet               output no diagnostic/information messages\n\
      --help                display this help and exit\n\
      --version             output version information and exit\n\
\n\
The block-size is fixed to 2048 bytes.\n\
\n\
All following arguments after --execute are taken to be arguments\n\
to the command to execute for each selected file. The string `{}'\n\
is replaced by the current file name being processed.\n\
\n\
Report bugs to <" BUG_EMAIL ">.\n"

static void print_knapsack_status(uint32_t current, uint32_t total);
static void verbose_printf(int minlvl, char *format, ...);
static void execute_command(char *file, char **argv);

static struct option long_options[] = {
    { "80min",	    	no_argument,	    NULL, '8' },
    { "74min",	    	no_argument,	    NULL, '7' },
    { "sldvd",	    	no_argument,	    NULL, '5' },
    { "dldvd",	    	no_argument,	    NULL, '9' },
    { "dirs-only",   	no_argument,	    NULL, 'd' },
    { "space",          required_argument,  NULL, 's' },
    { "reserve",    	required_argument,  NULL, 'r' },
    { "help",           no_argument,        NULL, 151 },
    { "version",        no_argument,        NULL, 150 },
    { "quiet",          no_argument,        NULL, 'q' },
    { "silent",         no_argument,        NULL, 'q' },
    { "invert",         no_argument,        NULL, 'i' },
    { "execute",        no_argument,        NULL, 'e' },
    { 0, 0, 0, 0 }
};

/* Print an error message to the standard error stream.
 */
static void
verbose_printf(int minlvl, char *format, ...)
{
    va_list args;

    va_start(args, format);
    if (verbosity >= minlvl) {
	vfprintf(stderr, format, args);
    	fflush(stderr);
    }
    va_end(args);
}

/* Called by knapsack function to display status information.
 */
static void
print_knapsack_status(uint32_t current, uint32_t total)
{
    verbose_printf(1, "  %4.1f%%\r", current*100.0/total);
}

/* Execute a command with specified arguments for the specified file.
 * Replace all occurences of {} in /argv/ with /file/.
 */
static void
execute_command(char *file, char **argv)
{
    pid_t child;

    child = fork();
    if (child == 0) {	/* in child */
    	int i, j, k;
	char *t;

    	/* Replace all occurences of {} with /file/ */
    	for (i = 1; argv[i] != NULL; i++) {
	    t = argv[i];
	    for (j = 0; (t = strstr(t, "{}")) != NULL; j++)
	    	t++;

    	    if (j > 0) {
    	    	t = xmalloc(strlen(argv[i]) - j*2 + j*strlen(file) + 1);
		k = 0;
	    	for (j = 0; argv[i][j] != 0; j++) {
		    if (argv[i][j] == '{' && argv[i][j+1] == '}') {
		    	strcpy(&t[k], file);
			k += strlen(file);
		    	j++;
		    } else {
		    	t[k++] = argv[i][j];
		    }
		}
		t[k] = '\0';
	    	argv[i] = t;
	    }
	}
    
    	execvp(argv[0], argv);
    	perror(program_name);
    	exit(1);
    } else if (child != -1) { /* in parent */
    	int status = 0;
    	if (waitpid(child, &status, 0) == -1) {
	    perror(program_name);
	    exit(1);
	}
	if (!WIFEXITED(status)) {
	    fprintf(stderr, "%s: child did not exit normally\n", program_name);
	    exit(1);
	}
    } else {
    	perror(program_name);
	exit(1);
    }
}

int
main (int argc, char **argv)
{
    int i;
    int j;
    int nonoptscount;
    uint64_t total;
    uint32_t used;
    uint32_t *weights;
    bool *results;
    char **nonopts;
    int space = ISO_BLOCKS_80_MIN_CD;
    int reserve = DEFAULT_RESERVED_BLOCKS;
    bool directories_only = FALSE;
    bool invert = FALSE;
    bool execute = FALSE;

    program_name = argv[0];
    nonopts = (argc <= 1 ? NULL : xmalloc(sizeof(char *) * (argc-1)));
    nonoptscount = 0;

    while (TRUE) {
    	int option_index = 0;
	i = getopt_long(argc, argv, "-8759dqver:s:", long_options, &option_index);
    	if (i == EOF)
	    break;

    	if (i == 'e') {
	    if (argc <= optind) {
	    	printf("%s: missing arguments for --execute\n", program_name);
		exit(1);
	    }
	    execute = TRUE;
	    break;
	}

    	switch (i) {
	case '8': /* --80min */
	    space = ISO_BLOCKS_80_MIN_CD;
	    break;

	case '7': /* --74min */
	    space = ISO_BLOCKS_74_MIN_CD;
	    break;
	
	case '5': /* --4,7 GB DVD */
	    space = ISO_BLOCKS_DVD_SL;
	    break;

	case '9': /* --8,5 GB DVD */
	    space = ISO_BLOCKS_DVD_DL;
	    break;

    	case 'd': /* --dirs-only */
	    directories_only = TRUE;
	    break;

    	case 'q': /* --quiet */
	    verbosity = 0;
	    break;

	case 'v': /* --verbose */
	    verbosity = 2;
	    break;

    	case 's': /* --space */
	    if (!xstrtoi(optarg, &space)) {
		fprintf(stderr, "%s: invalid integer: %s\n", program_name, optarg);
		exit(1);
	    }
    	    break;

    	case 'r': /* --reserve */
	    if (!xstrtoi(optarg, &reserve)) {
		fprintf(stderr, "%s: invalid integer: %s\n", program_name, optarg);
		exit(1);
	    }
    	    break;

    	case 'i': /* --invert */
	    invert = TRUE;
	    break;

	case 150: /* --version */
	    printf(VERSION_MESSAGE);
	    exit(0);

	case 151: /* --help */
	    printf(HELP_MESSAGE, program_name, DEFAULT_RESERVED_BLOCKS);
	    exit(0);

    	case 1: /* non-option arguments */
	    nonopts[nonoptscount++] = optarg;
	    break;

	case '?':
	default:
	    exit(1);
    	}
    }

    if (verbosity == 1 && !isatty(STDERR_FILENO))
    	verbosity = 0;

    if (nonoptscount == 0) {
	fprintf(stderr, "%s: missing arguments\n", program_name);
	fprintf(stderr, "Try `%s --help' for more information.\n", program_name);
	exit(1);
    }

    weights = xmalloc(sizeof(uint32_t) * nonoptscount);
    for (i = 0; i < nonoptscount; i++) {
    	uint64_t size;

	if (directories_only && !is_directory(nonopts[i])) {
	    verbose_printf(2, "ignored file `%s'\n", nonopts[i]);
    	    nonoptscount--;
    	    for (j = i; j < nonoptscount; j++)
	    	nonopts[j] = nonopts[j+1];
	    i--;
	    continue;
	}

        verbose_printf(2, "scanning `%s': ", nonopts[i]);
    	size = get_simple_file_size(nonopts[i], DEFAULT_BLOCK_SIZE);
	if (size == -1) {
	    verbose_printf(2, "failed\n");
	    fprintf(stderr, "%s: %s\n", get_error_file(), strerror(errno));
	    exit(1);
	}
    	verbose_printf(2, "%lld %s\n", size, (size == 1 ? "byte" : "bytes"));

    	if ((uint64_t) (size/DEFAULT_BLOCK_SIZE) > (uint64_t) UINT_MAX) {
	    verbose_printf(2, "`%s' has more than 2^32 block - ignoring\n", nonopts[i]);
    	    nonoptscount--;
    	    for (j = i; j < nonoptscount; j++)
	    	nonopts[j] = nonopts[j+1];
	    i--;
	    continue;
    	}

    	weights[i] = size/DEFAULT_BLOCK_SIZE;
    }

    if (nonoptscount == 0) {
	fprintf(stderr, "%s: no remaining arguments\n", program_name);
	exit(1);
    }

    space -= reserve;
    results = knapsack(weights, nonoptscount, space, print_knapsack_status,
    	    	    (verbosity == 0 ? INT_MAX : 10000));
    verbose_printf(1, "       \r");

    used = 0;
    total = 0;
    for (i = 0; i < nonoptscount; i++) {
    	if (invert ^ results[i]) {
	    if (execute) {
    	    	verbose_printf(2, "%s\n", nonopts[i]);
		execute_command(nonopts[i], &argv[optind]);
	    } else {
	    	printf("%s\n", nonopts[i]);
	    }
	    used += weights[i];
	} else if (invert) {
	    verbose_printf(2, "include `%s'\n", nonopts[i]);
	} else {
	    verbose_printf(2, "did not include `%s'\n", nonopts[i]);
	}
	total += weights[i];
    }

    verbose_printf(1, "used %d of %lld %s (%.1fM of %.1fM)\n",
    	    used,
    	    total,
	    (total == 1 ? "block" : "blocks"),
	    (used*DEFAULT_BLOCK_SIZE) / 1048576.0,
	    (total*DEFAULT_BLOCK_SIZE) / 1048576.0);

    verbose_printf(1, "wasted %d of %d %s (%.1fM of %.1fM)\n",
	    space-used,
	    space,
	    (space == 1 ? "block" : "blocks"),
	    (space-used)*DEFAULT_BLOCK_SIZE / 1048576.0,
	    (space*DEFAULT_BLOCK_SIZE) / 1048576.0);

    free(results);
    free(weights);
    free(nonopts);

    exit(0);
}
